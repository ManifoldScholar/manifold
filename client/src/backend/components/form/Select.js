import React, { Component } from "react";
import PropTypes from "prop-types";
import labelId from "helpers/labelId";
import GlobalForm from "global/components/form";
import Instructions from "./Instructions";
import withFormOptions from "hoc/with-form-options";
import IconComposer from "global/components/utility/IconComposer";

class FormSelect extends Component {
  static displayName = "Form.Select";

  static propTypes = {
    value: PropTypes.any,
    errors: PropTypes.array,
    label: PropTypes.string,
    instructions: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    name: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.any.isRequired,
        label: PropTypes.string.isRequired,
        internalValue: PropTypes.any.isRequired
      })
    ).isRequired,
    focusOnMount: PropTypes.bool,
    id: PropTypes.string,
    idForError: PropTypes.string
  };

  static defaultProps = {
    id: labelId("select-"),
    idForError: labelId("select-error-")
  };

  componentDidMount() {
    if (this.props.focusOnMount === true && this.inputElement) {
      this.inputElement.focus();
    }
  }

  render() {
    const options = this.props.options.map(option => {
      return (
        <option key={option.internalValue} value={option.internalValue}>
          {option.label}
        </option>
      );
    });

    return (
      <div className="form-input">
        <GlobalForm.Errorable
          className="form-input"
          name={this.props.name}
          errors={this.props.errors}
          label={this.props.label}
          idForError={this.props.idForError}
        >
          <label htmlFor={this.id}>{this.props.label}</label>
          <div className="form-select">
            <IconComposer
              icon="disclosureDown16"
              size={20}
              iconClass="form-select__icon"
            />
            <select
              id={this.id}
              aria-describedby={this.props.idForError}
              onChange={this.props.onChange}
              value={this.props.value}
              ref={input => {
                this.inputElement = input;
              }}
            >
              {options}
            </select>
          </div>
          <Instructions instructions={this.props.instructions} />
        </GlobalForm.Errorable>
      </div>
    );
  }
}

export default withFormOptions(FormSelect);
