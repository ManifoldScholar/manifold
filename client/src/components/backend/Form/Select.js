import React, { Component, PropTypes } from 'react';
import { Form as GlobalForm } from 'components/global';
import classNames from 'classnames';
import setter from './setter';

class FormSelect extends Component {

  static displayName = "Form.Select";

  static propTypes = {
    value: PropTypes.any,
    errors: PropTypes.array,
    label: PropTypes.string,
    name: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.any.isRequired,
      label: PropTypes.string.isRequired
    })).isRequired
  };

  render() {

    const options = this.props.options.map((option) => {
      return (
        <option key={option.value} value={option.value}>{option.label}</option>
      );
    });

    return (
      <div className="form-input">
        <GlobalForm.Errorable
          className="form-input"
          name={this.props.name}
          errors={this.props.errors}
          label={this.props.label}
        >
          <label>{this.props.label}</label>
          <div className="form-select">
            <i className="manicon manicon-caret-down"></i>
            <select
              onChange={this.props.onChange}
              value={this.props.value}
            >
              {options}
            </select>
          </div>
        </GlobalForm.Errorable>
      </div>
    );
  }

}

export default setter(FormSelect);
