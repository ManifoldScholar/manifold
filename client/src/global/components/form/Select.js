import React, { Component } from "react";
import PropTypes from "prop-types";
import { UID } from "react-uid";
import Errorable from "global/components/form/Errorable";
import Instructions from "./Instructions";
import withFormOptions from "hoc/with-form-options";
import IconComposer from "global/components/utility/IconComposer";
import classNames from "classnames";

class FormSelect extends Component {
  static displayName = "Form.Select";

  static propTypes = {
    value: PropTypes.any,
    errors: PropTypes.array,
    label: PropTypes.string,
    instructions: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    rounded: PropTypes.bool,
    name: PropTypes.string,
    wide: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.any.isRequired,
        label: PropTypes.string.isRequired,
        key: PropTypes.string.isRequired
      })
    ).isRequired,
    focusOnMount: PropTypes.bool
  };

  static defaultProps = {
    rounded: false
  };

  componentDidMount() {
    if (this.props.focusOnMount === true && this.inputElement) {
      this.inputElement.focus();
    }
  }

  get idPrefix() {
    return "select";
  }

  get idForErrorPrefix() {
    return "select-error";
  }

  get idForInstructionsPrefix() {
    return "select-instructions";
  }

  render() {
    const options = this.props.options.map(option => {
      return (
        <option key={option.key} value={option.value}>
          {option.label}
        </option>
      );
    });

    const selectClassName = classNames({
      "form-select": true,
      "form-select--rounded": this.props.rounded
    });

    const wrapperClassName = classNames({
      "form-input": true,
      wide: this.props.wide
    });

    const labelClassName = classNames({
      "pad-bottom": this.props.rounded
    });

    return (
      <UID>
        {id => (
          <div className={wrapperClassName}>
            <Errorable
              className={selectClassName}
              name={this.props.name}
              errors={this.props.errors}
              label={this.props.label}
              idForError={`${this.idForErrorPrefix}-${id}`}
            >
              <label
                htmlFor={`${this.idPrefix}-${id}`}
                className={labelClassName}
              >
                {this.props.label}
              </label>
              <div className="form-select">
                <IconComposer
                  icon="disclosureDown24"
                  size={24}
                  iconClass="form-select__icon"
                />
                <select
                  id={`${this.idPrefix}-${id}`}
                  aria-describedby={`${this.idForErrorPrefix}-${id} ${this.idForInstructionsPrefix}-${id}`}
                  onChange={this.props.onChange}
                  value={this.props.optionsMeta.stringValue}
                  ref={input => {
                    this.inputElement = input;
                  }}
                >
                  {options}
                </select>
              </div>
              <Instructions
                instructions={this.props.instructions}
                id={`${this.idForInstructionsPrefix}-${id}`}
              />
            </Errorable>
          </div>
        )}
      </UID>
    );
  }
}

export default withFormOptions(FormSelect);
