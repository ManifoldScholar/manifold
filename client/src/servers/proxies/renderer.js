import config from "config";
import proxy from "http-proxy-middleware";
import ch from "../../helpers/consoleHelpers";
import createStore from "../../store/createStore";
import exceptionRenderer from "../../helpers/exceptionRenderer";
import manifoldBootstrap from "bootstrap";
import CookieHelper from "helpers/cookie/Server";

const ssrRenderUrl = `http://${config.services.client.domain}:${config.services.client.sparePort}`;

export default function makeRendererProxy(requestHandler) {
  const reactServerProxy = proxy({
    target: ssrRenderUrl,
    changeOrigin: true,
    logLevel: "silent",

    onError: (err, req, res) => {
      ch.error(`Error proxying to port ${ssrRenderUrl}`);
      ch.error(`Perhaps the SSR service is in the process of reloading?`);
      ch.error(`Falling back to SSR rescue service only`);

      const store = createStore();

      const render = () => {
        try {
          requestHandler(req, res);
        } catch (error) {
          if (error.code === "MODULE_NOT_FOUND") {
            const msg =
              "Waiting for initial Webpack build to complete. Wait a few seconds " +
              "and reload.";
            ch.error(msg);
            res.setHeader("Content-Type", "text/html");
            res.end(msg);
          } else {
            ch.error(
              `Rendering fallback failed to render inn server-developmet.js`
            );
            res.setHeader("Content-Type", "text/html");
            res.end(exceptionRenderer(error));
          }
        }
      };

      const promises = [];
      const cookie = new CookieHelper(req, res);
      promises.push(manifoldBootstrap(store.getState, store.dispatch, cookie));
      Promise.all(promises).then(render, render);
    }
  });
  return reactServerProxy;
}
