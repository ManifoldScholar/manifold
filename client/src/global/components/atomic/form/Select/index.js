import React from "react";
import PropTypes from "prop-types";
import { useUID } from "react-uid";
import * as Styled from "./styles";

function Select({ label, value, options, onChange, preIcon }) {
  const uid = useUID();

  return (
    <Styled.Wrapper>
      <label htmlFor={uid} className="screen-reader-text">
        {label}
      </label>
      {preIcon && <Styled.PreIcon icon={preIcon} size={24} />}
      <Styled.Select id={uid} onChange={onChange} value={value}>
        {options.map(({ value: optValue, label: optLabel }) => (
          <option key={optLabel} value={optValue}>
            {optLabel}
          </option>
        ))}
      </Styled.Select>
      <Styled.DisclosureIcon icon="disclosureDown16" size={24} />
    </Styled.Wrapper>
  );
}

Select.displayName = "Global.Atomic.Form.Select";

export const selectShape = {
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired
    }).isRequired
  ).isRequired,
  onChange: PropTypes.func.isRequired,
  preIcon: PropTypes.string
};

Select.propTypes = selectShape;

export default Select;
