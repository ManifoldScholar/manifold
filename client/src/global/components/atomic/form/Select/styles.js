import styled from "@emotion/styled";
import IconComposer from "global/components/utility/IconComposer";
import {
  defaultTransitionProps,
  unstyledSelect,
  utilityPrimary
} from "theme/styles/mixins";
import { entityFilterForm } from "theme/styles/variables/crossComponent";

const ICON_WIDTH = 24;
const ICON_PADDING = 12;

export const borderStyles = `
  border: 1px solid var(--select-border-color);
  border-radius: 6px;
`;

export const Wrapper = styled.div`
  position: relative;
  flex-basis: ${entityFilterForm.selectMinWidth}px;
`;

const sharedIconStyles = `
  position: absolute;
  top: 50%;
  color: var(--select-border-color);
  pointer-events: none;
  transform: translateY(-50%);
`;

export const DisclosureIcon = styled(IconComposer)`
  ${sharedIconStyles}
  inset-inline-end: ${ICON_PADDING}px;
`;

export const PreIcon = styled(IconComposer)`
  ${sharedIconStyles}
  inset-inline-start: ${ICON_PADDING}px;
`;

export const Select = styled.select`
  ${unstyledSelect}
  ${utilityPrimary}
  padding: 11px ${ICON_WIDTH +
    ICON_PADDING * 1.5}px 13px var(--padding-inline-start, 13px);
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100%;
  font-size: 13px;
  ${borderStyles}
  transition: border-color ${defaultTransitionProps};

  &.focus-visible {
    border-color: var(--hover-color);
  }

  ${PreIcon} + & {
    --padding-inline-start: ${ICON_WIDTH + ICON_PADDING * 2}px;
  }

  option {
    color: var(--color-base-neutral-black);
  }
`;
