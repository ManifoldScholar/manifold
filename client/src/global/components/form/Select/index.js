import React, { Component } from "react";
import PropTypes from "prop-types";
import { UIDConsumer } from "react-uid";
import Errorable from "../Errorable";
import Instructions from "../Instructions";
import withFormOptions from "hoc/withFormOptions";
import BaseLabel from "../BaseLabel";
import { FormContext } from "helpers/contexts";
import * as Styled from "./styles";

export class FormSelect extends Component {
  static displayName = "Form.Select";

  static propTypes = {
    value: PropTypes.any,
    errors: PropTypes.array,
    label: PropTypes.string,
    hideLabel: PropTypes.bool,
    instructions: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    rounded: PropTypes.bool,
    name: PropTypes.string,
    wide: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.any.isRequired,
        label: PropTypes.string.isRequired,
        key: PropTypes.string.isRequired
      })
    ).isRequired,
    focusOnMount: PropTypes.bool
  };

  static defaultProps = {
    rounded: false,
    hideLabel: false
  };

  static contextType = FormContext;

  componentDidMount() {
    if (this.props.focusOnMount === true && this.inputElement) {
      this.inputElement.focus();
    }
  }

  get idPrefix() {
    return "select";
  }

  get idForErrorPrefix() {
    return "select-error";
  }

  get idForInstructionsPrefix() {
    return "select-instructions";
  }

  render() {
    const options = this.props.options.map(option => {
      return (
        <option key={option.key} value={option.value}>
          {option.label}
        </option>
      );
    });

    const styleType = this.context?.styleType;

    /* eslint-disable no-nested-ternary */
    const WrapperTag = this.props.rounded
      ? Styled.TertiarySelectWrapper
      : styleType === "primary"
      ? Styled.PrimarySelectWrapper
      : Styled.SecondarySelectWrapper;

    const SelectComponent = this.props.rounded
      ? Styled.TertiarySelect
      : styleType === "primary"
      ? Styled.PrimarySelect
      : Styled.SecondarySelect;

    const IconComponent = this.props.rounded
      ? Styled.IconTertiary
      : Styled.Icon;

    return (
      <UIDConsumer>
        {id => (
          <Errorable
            name={this.props.name}
            errors={this.props.errors}
            label={this.props.label}
            idForError={`${this.idForErrorPrefix}-${id}`}
          >
            <BaseLabel
              id={`${this.idPrefix}-${id}`}
              styleType={this.props.rounded ? "tertiary" : styleType}
              label={this.props.label}
              className={
                this.props.hideLabel ? "screen-reader-text" : undefined
              }
              isSelect
            />
            <WrapperTag $wide={this.props.wide}>
              <IconComponent icon="disclosureDown24" size={24} />
              <SelectComponent
                id={`${this.idPrefix}-${id}`}
                aria-describedby={`${this.idForErrorPrefix}-${id} ${this.idForInstructionsPrefix}-${id}`}
                onChange={this.props.onChange}
                value={
                  this.props.optionsMeta?.stringValue ?? this.props.value ?? ""
                }
                ref={input => {
                  this.inputElement = input;
                }}
                $wide={this.props.wide}
              >
                {options}
              </SelectComponent>
            </WrapperTag>
            {this.props.instructions && (
              <Instructions
                instructions={this.props.instructions}
                id={`${this.idForInstructionsPrefix}-${id}`}
                styleType={this.props.rounded ? "tertiary" : styleType}
                label={this.props.label}
                isSelect
                className={
                  this.props.hideLabel ? "screen-reader-text" : undefined
                }
              />
            )}
          </Errorable>
        )}
      </UIDConsumer>
    );
  }
}

export default withFormOptions(FormSelect);
