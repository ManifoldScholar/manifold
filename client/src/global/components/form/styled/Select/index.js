import React, { Component } from "react";
import PropTypes from "prop-types";
import { UIDConsumer } from "react-uid";
import Errorable from "../Errorable";
import Instructions from "../Instructions";
import withFormOptions from "hoc/withFormOptions";
import BaseLabel from "../BaseLabel";
import FieldWrapper from "../FieldWrapper";
import { FormContext } from "helpers/contexts";
import * as Styled from "./styles";

class FormSelect extends Component {
  static displayName = "Form.Select";

  static propTypes = {
    value: PropTypes.any,
    errors: PropTypes.array,
    label: PropTypes.string,
    instructions: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    rounded: PropTypes.bool,
    name: PropTypes.string,
    wide: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.any.isRequired,
        label: PropTypes.string.isRequired,
        key: PropTypes.string.isRequired
      })
    ).isRequired,
    focusOnMount: PropTypes.bool
  };

  static defaultProps = {
    rounded: false
  };

  static contextType = FormContext;

  componentDidMount() {
    if (this.props.focusOnMount === true && this.inputElement) {
      this.inputElement.focus();
    }
  }

  get idPrefix() {
    return "select";
  }

  get idForErrorPrefix() {
    return "select-error";
  }

  get idForInstructionsPrefix() {
    return "select-instructions";
  }

  render() {
    const options = this.props.options.map(option => {
      return (
        <option key={option.key} value={option.value}>
          {option.label}
        </option>
      );
    });

    const styleType = this.context.styleType;

    const WrapperTag =
      styleType === "primary" ? Styled.PrimarySelectWrapper : "div";

    const SelectComponent = this.props.rounded
      ? Styled.TertiarySelect
      : styleType === "primary"
      ? Styled.PrimarySelect
      : Styled.SecondarySelect;

    const IconComponent = this.props.rounded
      ? Styled.IconTertiary
      : Styled.Icon;

    return (
      <UIDConsumer>
        {id => (
          <FieldWrapper className="wide">
            <Errorable
              name={this.props.name}
              errors={this.props.errors}
              label={this.props.label}
              idForError={`${this.idForErrorPrefix}-${id}`}
            >
              <BaseLabel
                id={`${this.idPrefix}-${id}`}
                styleType={this.props.rounded ? "tertiary" : styleType}
                label={this.props.label}
                isSelect
              />
              <WrapperTag>
                <IconComponent icon="disclosureDown24" size={24} />
                <SelectComponent
                  id={`${this.idPrefix}-${id}`}
                  aria-describedby={`${this.idForErrorPrefix}-${id} ${this.idForInstructionsPrefix}-${id}`}
                  onChange={this.props.onChange}
                  value={this.props.optionsMeta.stringValue}
                  ref={input => {
                    this.inputElement = input;
                  }}
                >
                  {options}
                </SelectComponent>
              </WrapperTag>
              <Instructions
                instructions={this.props.instructions}
                id={`${this.idForInstructionsPrefix}-${id}`}
              />
            </Errorable>
          </FieldWrapper>
        )}
      </UIDConsumer>
    );
  }
}

export default withFormOptions(FormSelect);
