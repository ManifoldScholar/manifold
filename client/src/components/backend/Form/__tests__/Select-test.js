import React from 'react';
import renderer from 'react-test-renderer';
import Select from '../Select';

describe("Backend.Form.Select component", () => {

  const changeMock = jest.fn();
  const options = [
    {
      label: "option-1",
      value: "1"
    },
    {
      label: "option-2",
      value: "2"
    }
  ];

  it('renders correctly', () => {
    const component = renderer.create(
      <Select
        options={options}
        label="Label this"
        name="attributes[fake]"
        onChange={changeMock}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
