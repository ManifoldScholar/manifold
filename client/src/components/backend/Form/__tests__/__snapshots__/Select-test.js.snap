// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend.Form.Select component renders correctly 1`] = `
<div
  className="form-input"
>
  <div
    className="form-input"
    style={Object {}}
  >
    <label
      htmlFor={undefined}
    >
      Label this
    </label>
    <div
      className="form-select"
    >
      <i
        aria-hidden="true"
        className="manicon manicon-caret-down"
      />
      <select
        aria-describedby="select-error-24"
        id={undefined}
        onChange={[Function]}
        value={undefined}
      >
        <option
          value="1"
        >
          option-1
        </option>
        <option
          value="2"
        >
          option-2
        </option>
      </select>
    </div>
  </div>
</div>
`;
